### DISCUSSION TITLE
Metaclass method meets reportSelfClsParameterName

### DISCUSSION THREAD

**Discussion by AsgerJon** [microsoft/pyright#7442] [General]
Instance methods in a custom metaclass are allowed to use `cls` instead of `self` by the `reportSelfClsParameterName` rule. If the custom metaclass decorates a method such as `__prepare__`, with the `@classmethod` decorator, the rule insists that the first argument be `cls`. However, the rule does allow the first argument in a custom implementation of `__new__` in the metaclass to be named `mcls`. (Never explicitly decorate the `__new__` method in a custom metaclass with `@classmethod`! You are gonna have a bad time if you do).

Custom metaclasses are not for the faint of heart. Currently, the `reportSelfClsParameterName` rule adds to the confusion by allowing `self` and `cls` as the first argument in instance methods, allowing `mcls` in the `__new__` class method, but disallows other class methods from using `mcls` as the first argument.

**Comment by erictraut** [comment]
Arguably, this rule doesn't belong in a type checker, as it's more of a linter (stylistic) rule. I added it very early in the evolution of pyright. I've considered removing it entirely, but some users seem to like it. If it doesn't suit your needs, you can disable it in your project.
